### PR TITLE
Bg 66934 sign p2tr musig2

### PR DESCRIPTION
**Pull request by saravanan7mani**

### CODE REVIEW

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts [hunk 1]
```diff
@@ -122,6 +134,20 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint> = UtxoTransaction<bigin
...
+  private getOriginalMusig2PubNonce(participantPubKey: Buffer, tapOutputKey: Buffer): Uint8Array | undefined {
+    return this.MUSIG2_PUB_NONCE_CACHE.get(this.createMusig2PubNonceCacheKey(participantPubKey, tapOutputKey));
+  }
+  private setOriginalMusig2PubNonce(participantPubKey: Buffer, tapOutputKey: Buffer, pubNonce: Uint8Array): void {
+    this.MUSIG2_PUB_NONCE_CACHE.set(this.createMusig2PubNonceCacheKey(participantPubKey, tapOutputKey), pubNonce);
+  }
+  private deleteOriginalMusig2PubNonce(participantPubKey: Buffer, tapOutputKey: Buffer): void {
+    this.MUSIG2_PUB_NONCE_CACHE.delete(this.createMusig2PubNonceCacheKey(participantPubKey, tapOutputKey));
+  }
+
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/UtxoPsbt.ts:180]: see commits with NonceStore

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts [hunk 2]
```diff
@@ -595,6 +650,99 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint> = UtxoTransaction<bigin
...
+    return { tapOutputKey, participantPubKey, pubNonce };
+  }
+
+  /**
+   * Generates and sets Musig2 nonces to p2trMusig2 key path spending inputs.
+   * tapInternalkey, tapMerkleRoot, tapBip32Derivation for rootWalletKey are required per p2trMusig2 key path input.
+   * Also participant keys are required from psbt proprietary key values.
+   * Ref: https://gist.github.com/sanket1729/4b525c6049f4d9e034d27368c49f28a6
+   * @param psbt
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/UtxoPsbt.ts:845]: :scissors:

**Review by OttoAllmendinger — CHANGES_REQUESTED** [review]

File: modules/utxo-lib/src/noble_ecc.ts [hunk 3]
```diff
@@ -40,7 +40,7 @@ function isPoint(p: Uint8Array, xOnly: boolean): boolean {
 }
 
 function toBigInt(b: Uint8Array | Buffer): bigint {
-  return Buffer.from(b).readBigUint64BE();
+  return BigInt(`0x${Buffer.from(b).toString('hex')}`);
```
[the remark this answers is not in the document]
Comment (saravanan7mani) [reply modules/utxo-lib/src/noble_ecc.ts]: Done

File: modules/utxo-lib/src/bitgo/Musig2.ts [hunk 4]
```diff
@@ -100,129 +97,69 @@ export function decodePsbtMusig2ParticipantsKeyValData(
...
-
-  if (myDerivation.length !== 1) {
-    throw new Error('root wallet key should derive one tapBip32Derivation');
+export function createMusig2Nonce(
+  privateKey: Uint8Array,
+  publicKey: Uint8Array,
+  xOnlyPublicKey: Uint8Array,
+  txHash: Uint8Array,
+  sessionId?: Buffer
+): Uint8Array {
```
[the remark this answers is not in the document]
Comment (saravanan7mani) [reply modules/utxo-lib/src/bitgo/Musig2.ts]: Done

File: modules/utxo-lib/src/bitgo/Musig2.ts [hunk 5]
```diff
@@ -100,129 +97,69 @@ export function decodePsbtMusig2ParticipantsKeyValData(
...
-  if (!input.tapMerkleRoot) {
-    throw new Error('tapMerkleRoot is required to generate nonce');
-  }
-
-  if (!input.tapBip32Derivation?.length) {
-    throw new Error('tapBip32Derivation is required to generate nonce');
-  }
-
+  inputIndex: number
+): PsbtMusig2ParticipantsKeyValueData | undefined {
```
[the remark this answers is not in the document]
Comment (saravanan7mani) [reply modules/utxo-lib/src/bitgo/Musig2.ts:252]: Done

File: modules/utxo-lib/src/bitgo/Musig2.ts [hunk 6]
```diff
@@ -100,129 +97,69 @@ export function decodePsbtMusig2ParticipantsKeyValData(
...
-  if (!tapOutputKey.equals(participantKeyValData.tapOutputKey)) {
-    throw new Error('Invalid participants keyata tapOutputKey');
-  }
+  return decodePsbtMusig2ParticipantsKeyValData(participantsKeyVals[0]);
+}
 
-  if (!tapInternalKey.equals(input.tapInternalKey)) {
-    throw new Error('tapInternalKey and aggregated participant pub keys does not match');
-  }
+export function validatePsbtMusig2ParticipantsKeyValData(
```
[the remark this answers is not in the document]
Comment (saravanan7mani) [reply modules/utxo-lib/src/bitgo/Musig2.ts]: Done

File: modules/utxo-lib/src/taproot.ts [hunk 7]
```diff
@@ -115,7 +115,10 @@ export function hashTapBranch(child1: Buffer, child2: Buffer): Buffer {
   return bcrypto.taggedHash('TapBranch', Buffer.concat(sortedChildren));
 }
 
-function calculateTapTweak(pubkey: Uint8Array, taptreeRoot?: Uint8Array): Uint8Array {
+export function calculateTapTweak(pubkey: Uint8Array, taptreeRoot?: Uint8Array): Uint8Array {
+  if (!pubkey.length) {
```
[the remark this answers is not in the document]
Comment (saravanan7mani) [reply modules/utxo-lib/src/taproot.ts]: Done

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts (hunk 1 above)
Comment (saravanan7mani) [reply modules/utxo-lib/src/bitgo/UtxoPsbt.ts]: Done

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts (hunk 2 above)
Comment (saravanan7mani) [reply modules/utxo-lib/src/bitgo/UtxoPsbt.ts:838]: Done

File: modules/utxo-lib/src/bitgo/Musig2.ts [hunk 8]
```diff
@@ -22,6 +29,62 @@ export interface PsbtMusig2PubNonceKeyValueData {
...
+    for (const b of this.nonces) {
+      if (Buffer.from(b).equals(nonce)) {
+        return b;
+      }
+    }
+    throw new Error(`unknown nonce`);
+  }
+
+  /**
+   * Creates musig2 nonce.
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/Musig2.ts:62]: `Creates musig2 nonce and stores buffer reference.`

File: modules/utxo-lib/src/bitgo/Musig2.ts [hunk 9]
```diff
@@ -106,19 +209,43 @@ export function createTapOutputKey(internalPubKey: Buffer, tapTreeRoot: Buffer):
...
+    musig.partialSign({
+      secretKey: privateKey,
+      publicNonce: nonceStore.getRef(publicNonce),
+      sessionKey,
+    })
+  );
+}
+
+/**
+ * @returns parse psbt proprietary key for musig2 participant key value data
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/Musig2.ts:246]: `@returns psbt proprietary key...`

File: modules/utxo-lib/src/bitgo/Musig2.ts [hunk 10]
```diff
@@ -99,130 +177,180 @@ export function decodePsbtMusig2ParticipantsKeyValData(
...
-  const participantPubKeys = participantKeyValData.participantPubKeys;
+  return decodePsbtMusig2ParticipantsKeyValData(participantsKeyVals[0]);
+}
 
-  const tapInternalKey = createTapInternalKey(participantPubKeys);
-  if (!tapInternalKey.equals(participantKeyValData.tapInternalKey)) {
-    throw new Error('Invalid participants keyata tapInternalKey');
-  }
+/**
+ * @returns parse psbt proprietary key for musig2 public nonce key value data
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/Musig2.ts:270]:
```suggestion
 * @returns psbt proprietary key for musig2 public nonce key value data
```

File: modules/utxo-lib/src/bitgo/Musig2.ts [hunk 11]
```diff
@@ -99,130 +177,180 @@ export function decodePsbtMusig2ParticipantsKeyValData(
...
-  const derivedPubKey = deriveWalletPubKey(input.tapBip32Derivation, rootWalletKey);
-  const participantPubKey = participantPubKeys.find((pubKey) => pubKey.equals(derivedPubKey));
+  return nonceKeyVals.map((kv) => decodePsbtMusig2NonceKeyValData(kv));
+}
 
-  if (!Buffer.isBuffer(participantPubKey)) {
-    throw new Error('participant plain pub key should match one tapBip32Derivation plain pub key');
-  }
+/**
+ * @returns assert musig2 participant key value data with tapInternalKey and tapMerkleRoot.
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/Musig2.ts:294]:
```suggestion
 * Assert musig2 participant key value data with tapInternalKey and tapMerkleRoot.
```

**Review by OttoAllmendinger — CHANGES_REQUESTED** [review]
last round of nits

**Review by OttoAllmendinger — APPROVED** [review]

**Review by byhow — APPROVED** [review]

**Pull request merged**